### ISSUE TITLE
Illegal memory access for BC computation on as-Skitter

### ISSUE THREAD

**Issue by Adam27X** [gunrock/gunrock#17]
Hi all,

I'm trying to run some Betweenness Centrality tests using the following graph: http://www.cise.ufl.edu/research/sparse/matrices/SNAP/as-Skitter.html

Here's how I'm running the program:

$ ./bin/test_bc_7.0_x86_64 market ~/github/graph-utils/graphs/as-Skitter.mtx --device=0 --src=-1 --quick

and here's the output up until the first error:

  Reading directly from previously stored CSR arrays ...
Done reading (3s).

Degree Histogram (1696415 vertices, 22190596 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 218399 (12.87%)
    Degree 2^1: 447171 (26.36%)
    Degree 2^2: 456022 (26.88%)
    Degree 2^3: 318795 (18.79%)
    Degree 2^4: 149829 (8.83%)
    Degree 2^5: 68987 (4.07%)
    Degree 2^6: 21819 (1.29%)
    Degree 2^7: 9364 (0.55%)
    Degree 2^8: 3714 (0.22%)
    Degree 2^9: 1396 (0.08%)
    Degree 2^10: 536 (0.03%)
    Degree 2^11: 211 (0.01%)
    Degree 2^12: 85 (0.01%)
    Degree 2^13: 60 (0.00%)
    Degree 2^14: 24 (0.00%)
    Degree 2^15: 3 (0.00%)

[../../gunrock/app/bc/bc_enactor.cuh, 299] edge_map_forward::Kernel failed (CUDA error 77: an illegal memory access was encountered)

I think the problem comes from a specific source vertex, because the computation will work if I set src=5, for example. For what it's worth, I ran this using a Tesla K40c GPU and the CUDA 7.0 toolkit on Ubuntu 14.04.

If you need any more info from my end I'm happy to provide it.

Thanks!

-Adam

**Comment by yzhwang** [comment]
Hi Adam!
Thanks for the report. Currently Gunrock team are working hard for two papers (SC and IISWC) due on April 17th. Will look into this problem as soon as we pass the deadline.

**Comment by Adam27X** [comment]
Yangzihao, 

No worries. Good luck with your submissions!

-Adam

**Comment by yzhwang** [comment]
Hi Adam,

Sorry for the delay. I'm looking into this right now. Will keep you posted.

**Comment by yzhwang** [comment]
Hi Adam,

Hopefully I've fixed the bug in the latest commit. Since I'm adding a new operator which can do neighborlist reduce and advance in one pass. To use both exclusive/inclusive scan result on one array I introduced one offset but forgot to enlarge that array size by one too. Here's the result for as-Skitter graph:

bin/test_bc_7.0_x86_64 market /data/gunrock_dataset/large/as-Skitter/as-Skitter.mtx --device=3 --src=-1 --quick | tee result
  Reading directly from previously stored CSR arrays ...
Done reading (3s).

Degree Histogram (1696415 vertices, 22190596 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 218399 (12.87%)
    Degree 2^1: 447171 (26.36%)
    Degree 2^2: 456022 (26.88%)
    Degree 2^3: 318795 (18.79%)
    Degree 2^4: 149829 (8.83%)
    Degree 2^5: 68987 (4.07%)
    Degree 2^6: 21819 (1.29%)
    Degree 2^7: 9364 (0.55%)
    Degree 2^8: 3714 (0.22%)
    Degree 2^9: 1396 (0.08%)
    Degree 2^10: 536 (0.03%)
    Degree 2^11: 211 (0.01%)
    Degree 2^12: 85 (0.01%)
    Degree 2^13: 60 (0.00%)
    Degree 2^14: 24 (0.00%)
    Degree 2^15: 3 (0.00%)

  
GPU BC finished in 88868416.000000 msec.

It's like 24 hours and 52msec per node, which is not super good. We need to improve the performance! I'm trying to think about how to do multi-source BFS in Gunrock like what you did. Anyway, lemme know if it works.
